# HG changeset patch
# User Jaroslav Hajek <highegg@gmail.com>
# Date 1236286259 -3600
#      Thu Mar 05 21:50:59 2009 +0100
# Node ID 354179c24c79dcd00c2d14ca318661f0ba0d0929
# Parent  35cd375d4bb3c3d577bd3f01f1bbc25005315e02
fix hdf5 saving of diag & perm matrices

diff --git a/src/ChangeLog b/src/ChangeLog
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,8 +1,12 @@
+2009-03-05  Jaroslav Hajek  <highegg@gmail.com>
+	
+	* ls-hdf5.cc (add_hdf5_data): Handle diag & perm matrices.
+
 2009-03-05  John W. Eaton  <jwe@octave.org>
 
 	* pt-cell.cc, pt-cell.h (tree_cell::dup): Now const.
 	* comment-list.cc, comment-list.h (octave_comment_list::dup): Ditto.
 	* pt-arg-list.cc, pt-arg-list.h (tree_argument_list::dup): Ditto.
 	* pt-assign.cc, pt-assign.h (tree_simple_assignment::dup,
 	tree_multi_assignment::dup): Ditto.
 	* pt-binop.cc, pt-binop.h (tree_binary_expression::dup,
diff --git a/src/ls-hdf5.cc b/src/ls-hdf5.cc
--- a/src/ls-hdf5.cc
+++ b/src/ls-hdf5.cc
@@ -744,17 +744,17 @@ add_hdf5_data (hid_t loc_id, const octav
   hsize_t dims[3];
   hid_t type_id = -1, space_id = -1, data_id = -1, data_type_id = -1;
   bool retval = false;
   octave_value val = tc;
   // FIXME: diagonal & permutation matrices currently don't know how to save
   // themselves, so we convert them first to normal matrices using A = A(:,:).
   // This is a temporary hack.
   if (val.is_diag_matrix () || val.is_perm_matrix ())
-    val = val.do_index_op (octave_value_list (2, octave_value::magic_colon_t));
+    val = val.full_value ();
 
   std::string t = val.type_name();
 
   data_id = H5Gcreate (loc_id, name.c_str (), 0);
   if (data_id < 0)
     goto error_cleanup;
 
   // attach the type of the variable
